qiskit/pulse/frame.py

```diff
@@ -0,0 +1,216 @@
+        to allow for parametric identifiers in frames we separate the string prefix from
+        the numeric (and possibly parametric) index. This behaviour may change in the
+        future if parameters can be assigned string values.
+
+        Args:
+            prefix: The index of the frame.
+            index: An optional parameter to specify the numeric part of the index.
+
+        Raises:
+            PulseError: if the frame identifier is not a string or if the identifier
+                contains numbers which should have been specified as the index.
+        """
+        if index is not None:
+            validate_index(index)
+
+        if not isinstance(prefix, str):
+            raise PulseError(f"Frame identifiers must be string. Got {type(prefix)}.")
+
+        if any(char.isdigit() for char in prefix):
+            raise PulseError(f"Frame prefixes cannot contain digits. Found {prefix}")
+
+        self._identifier = (prefix, index)
+        self._hash = hash((type(self), self._identifier))
+
+    @property
+    def identifier(self) -> Tuple:
+        """Return the index of this frame. The index is a label for a frame."""
+        return self._identifier
+
+    @property
+    def prefix(self) -> str:
+        """Return the prefix of the frame."""
+        return self._identifier[0]
+
+    @property
+    def index(self) -> Union[int, ParameterExpression]:
+        """Return the index of the frame."""
+        return self._identifier[1]
+
+    @property
+    def name(self) -> str:
+        """Return the shorthand alias for this frame, which is based on its type and index."""
+        if isinstance(self._identifier[1], Parameter):
+            return f"{self._identifier[0]}{self._identifier[1].name}"
+
+        return f"{self._identifier[0]}{self._identifier[1]}"
+
+    @property
+    def parameters(self) -> Set:
+        """Parameters which determine the frame index."""
+        if isinstance(self._identifier[1], ParameterExpression):
+            return self._identifier[1].parameters
+
+        return set()
+
+    def is_parameterized(self) -> bool:
+        """Return true if the identifier has a parameter."""
+        return isinstance(self._identifier[1], ParameterExpression)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.name})"
+
+    def __eq__(self, other: "Frame") -> bool:
+        """Return True iff self and other are equal, specifically, iff they have the same type
+        and the same index.
+
+        Args:
+            other: The frame to compare to this frame.
+
+        Returns:
+            True iff equal.
+        """
+        return type(self) is type(other) and self._identifier == other._identifier
+
+    def __hash__(self):
+        return self._hash
+
+
+@dataclass
+class FrameDefinition:
+    """A class to keep track of frame definitions."""
+
+    # The frequency of the frame at time zero.
```

Comment:
These docstrings can be converted to attribute docstrings by moving after the attribute declaration. https://www.python.org/dev/peps/pep-0224/